### ISSUE TITLE
If decimal separator is set to "," shipping cost do not accept value like 5,5 € (round to 5,00) 

### ISSUE THREAD

**Issue by giodc** [woothemes/woocommerce#42955]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

If decimal separator is set to "," (comma - europe standard) when set a shipping cost that have a decimal (example: 5,5) woocommerce round to 5. 

tested locally on fresh woocommerce install, Version 8.4.0

### Expected behavior

possibility to set shipping cost with decimal like: 5,5 

### Actual behavior

WooCommerce round that to 5, if entered any value with decimal example: 5,5

### Steps to reproduce

1: Go to Settings > General
2: set Thousand separator to .
3: set Decimal separator to , 
4: set Number decimal to 2
5: Go to Settings > Shipping > edit a zone > add a flat rate with a decimal value

setting screenshot
![Screenshot 2023-12-19 at 7 55 03 a m](https://github.com/woocommerce/woocommerce/assets/2741044/cac2916a-6f8c-469c-b739-c37768c623a1)

shipping before save
![Screenshot 2023-12-19 at 7 56 37 a m](https://github.com/woocommerce/woocommerce/assets/2741044/1c449f24-213a-47da-9e79-73b82acb7cf4)

shipping after save
![Screenshot 2023-12-19 at 7 56 42 a m](https://github.com/woocommerce/woocommerce/assets/2741044/bc303529-6d38-42df-8744-bb5793b926b8)

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://localhost:10019
Site address (URL): http://localhost:10019
WC Version: 8.4.0
REST API Version: ✔ 8.4.0
WC Blocks Version: ✔ 11.6.2
Action Scheduler Version: ✔ 3.7.0
Log Directory Writable: ✔
WP Version: 6.4.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.1.23
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.1.2
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.4.0
WC Database Prefix: wp_
Total Database Size: 4.81MB
Database Data Size: 3.12MB
Database Index Size: 1.69MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.03MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (3) ###

All-in-One WP Migration S3 Extension: by ServMask – 3.77
All-in-One WP Migration: by ServMask – 7.76
WooCommerce: by Automattic – 8.4.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/
Checkout: #9 - /checkout/
My account: #10 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Three
Version: 1.1 (update to version 1.3 is available)
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: /Users/giovannidecarlo/Local Sites/woocommerce/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/templates/notices/error.php
/Users/giovannidecarlo/Local Sites/woocommerce/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/templates/notices/notice.php
/Users/giovannidecarlo/Local Sites/woocommerce/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/templates/notices/success.php

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-external-affiliate
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-12-20 13:44:20 +00:00
Options: ✔
Notes: 62
Onboarding: skipped

### Action Scheduler ###

Complete: 2
Oldest: 2023-12-19 13:45:21 +0000
Newest: 2023-12-19 13:45:21 +0000

Pending: 1
Oldest: 2023-12-20 13:45:21 +0000
Newest: 2023-12-20 13:45:21 +0000

### Status report information ###

Generated at: 2023-12-19 13:58:42 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by Simbaclaws** [comment]
I'm experiencing the same issue, has there been any workarounds so far?

**Comment by Seventhose** [comment]
Same issue here. I hope it can be resolved quickly.

**Comment by melek** [comment]
I did a little experimentation and found that the `5,50` value is being stored correctly, but not fetched correctly. Here is the value after changing it in Woo 8.4 and reverting to Woo 7.9:

![image](https://github.com/woocommerce/woocommerce/assets/220149/29339bc9-c213-4e7b-ad4e-3b7b07cf33a4)

I re-upgraded to Woo 8.4, and changed it to `5,01`, and again got the `5,00` display in Settings, but in the cart the correct amount was shown:
![image](https://github.com/woocommerce/woocommerce/assets/220149/2380a578-e5c4-4ae3-85aa-9ba38f044037)

Tried again with `5,51` and got the same result - it is saved correctly. This is confirmed by updating the value and checking the `wp_options` table for the method's entry:
![image](https://github.com/woocommerce/woocommerce/assets/220149/617c8069-3dd5-499f-a850-3d586d552076)

So at least in my case, the problem is a visual error when the `#woocommerce_flat_rate_cost` input is populated via AJAX in the shipping settings. I couldn't figure out what function exactly is called to render this value, but since it isn't present in older versions of Woo it was probably introduced along with the redesign of the built-in shipping method modals.

In other words, until it is fixed be aware that the value is likely getting saved and doesn't have to impact your pricing - just double check your front-end after saving to verify!

**Comment by nikman68** [comment]
Same issue here major problem.

**Comment by Toxdg** [comment]
Deletion/disabling WooPayments ? 
In my case it's helped

**Comment by nikman68** [comment]
In my case dont.I come back 8.3.1

**Comment by danielcharrua** [comment]
Confirmed issue in WooCommerce 8.4.0, the data gets stored correctly but once you open again the modal for editing the price, it gets rounded... Something is wrong...

**Comment by PunchcardCoder** [comment]
Same issue here. IMHO, the value is stored incorrectly: nowhere else is a decimal value stored with a decimal comma, storage in the database is always with a decimal point. Looks like a bug, smells a bug, it might even be a bug

**Comment by PunchcardCoder** [comment]
If I understand the workings of wc correctly (big if), the error is in WC_Shipping_Flat_Rate::sanitize_cost. It seems, this function replaces wc_get_price_thousand_separator() with '' at a time when the value already has "." as decimal point, thereby effectively removing the decimal point if your thousand-separator is ".". Any decimal amount becomes up to a factor 100 larger. 
Until this issue gets resolved, the following is a quick and dirty patch:
```
add_filter( 'woocommerce_shipping_flat_rate_instance_settings_values', 'one_helluva_patch_34567', 20, 2 );
function one_helluva_patch_34567($setting, $obj){
  if(isset($setting['cost'])){ 
    $setting['cost'] = preg_replace('/\D/', '.', sanitize_text_field($setting['cost']), 1); 
  }
  return $setting;
}
```

**Comment by adrianduffell** [comment]
I'm not currently able to reproduce this. Is anyone still experiencing issues with this?

**Comment by jessepearson** [comment]
Closing this out, as it is no longer able to be reproduced. I tested with WooCommerce 9.7.1 by setting my decimal points to be a comma and set my flat rate to 5,01, saved, refreshed, and when opening flat rate again, I get the proper value.
